### PR TITLE
Exchange v3 related updates

### PR DESCRIPTION
**Pull request by ovidiuolteanu**

### CODE REVIEW

File: contracts/locked_asset_contract.py [hunk 1]
```diff
@@ -76,6 +76,18 @@ def contract_upgrade(self, deployer: Account, proxy: ProxyNetworkProvider, bytec
 
         return upgrade_call(type(self).__name__, proxy, gas_limit, deployer, Address(self.address),
                             bytecode_path, metadata, arguments)
+    
+    def unlock_assets(self, user: Account, proxy: ProxyNetworkProvider, args: list):
+        """ Expected as args:
+            type[List[ESDTToken]]: tokens list
+        """
+        function_purpose = "unlock tokens early"
```
Comment (claudiulataretu) [inline contracts/locked_asset_contract.py:84]: This is not early unlock

File: contracts/metastaking_contract.py [hunk 2]
```diff
@@ -83,6 +83,9 @@ def contract_deploy(self, deployer: Account, proxy: ProxyNetworkProvider, byteco
 
     def contract_upgrade(self, deployer: Account, proxy: ProxyNetworkProvider, bytecode_path,
                          args: list = [], no_init: bool = False):
+        """ Expected as args:
```
Comment (claudiulataretu) [inline contracts/metastaking_contract.py:86]: metastaking upgrade doesn't require any arguments

File: events/event_generators.py [hunk 3]
```diff
@@ -475,8 +475,11 @@ def generateExitMetastakeEvent(context: Context, user: Account, metastake_contra
 
         decoded_metastake_tk_attributes = get_lp_from_metastake_token_attributes(metastake_token_attributes)
 
-        farm_tk_details = context.network_provider.api.get_non_fungible_token(
-            metastake_contract.farm_token, decoded_metastake_tk_attributes['lp_farm_token_nonce']
+        # farm_tk_details = context.network_provider.api.get_non_fungible_token(
```
Comment (claudiulataretu) [inline events/event_generators.py:478]: maybe remove commented code

File: tools/notebooks/boosted-farm.ipynb [hunk 4]
```diff
@@ -14,9 +14,197 @@
...
+     "output_type": "stream",
+     "text": [
+      "loaded 10 accounts from 1 PEM files.\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\u001b[38;20m2024-01-08 14:46:30,691 - DEBUG - Account.sync_nonce() done: 324\u001b[0m\n",
```
Comment (claudiulataretu) [inline tools/notebooks/boosted-farm.ipynb:40]: clear the outputs before committing any changes for notebooks

**Review by claudiulataretu — CHANGES_REQUESTED** [review]

File: contracts/dex_proxy_contract.py [hunk 5]
```diff
@@ -377,3 +458,60 @@ def print_contract_info(self):
...
+
+    def get_all_decoded_farm_token_attributes_from_network(self, api: ApiNetworkProvider, farm_token_nonce: int):
+        # Get token details for a given farm token
+        farm_token_on_network = api.get_non_fungible_token(self.proxy_farm_token, farm_token_nonce)
+
+        # Decode the farm token attributes
+        decoded_xmex_farm_attributes = decode_merged_attributes(base64_to_hex(farm_token_on_network.attributes), XMEXFARM_ATTRIBUTES)
+        logger.debug(decoded_xmex_farm_attributes)
+
+        # Decode the LP token attributes
```
[the remark this answers is not in the document]
Comment (ovidiuolteanu) [reply contracts/dex_proxy_contract.py:470]: Good point. I added `get_all_decoded_lp_token_attributes_from_network` later and didn't looked back to optimise the farm one and reduce duplicate code.
I'll modify.

File: tools/notebooks/boosted-farm.ipynb (hunk 4 above)
Comment (ovidiuolteanu) [reply tools/notebooks/boosted-farm.ipynb:40]: good catch. I usually do but this one escaped somehow.

File: events/event_generators.py (hunk 3 above)
Comment (ovidiuolteanu) [reply events/event_generators.py:478]: I wasn't sure whether I will have to get back to this and still am a bit. Proxy requires the address of an user holding the tokens to just read the token attributes so I had to do some changes to accommodate this and I don't really like it. Ideally I should've been able to use it as in API without necessarily passing an user, but will have to see.

Comment (ovidiuolteanu) [reply events/event_generators.py:478]: I will probably add a todo comment with this note, to know why I did this.

File: contracts/metastaking_contract.py (hunk 2 above)
Comment (ovidiuolteanu) [reply contracts/metastaking_contract.py:86]: Legacy stuff, mostly irrelevant now when forcing `no_init = True`.

File: contracts/fees_collector_contract.py [hunk 6]
```diff
@@ -49,6 +49,35 @@ def contract_deploy(self, deployer: Account, proxy: ProxyNetworkProvider, byteco
         tx_hash, address = deploy(type(self).__name__, proxy, gas_limit, deployer, bytecode_path, metadata, arguments)
 
         return tx_hash, address
+    
+    def contract_upgrade(self, deployer: Account, proxy: ProxyNetworkProvider, bytecode_path, args: list = None,
+                         no_init: bool = False):
+        """ Expected as args:
```
[the remark this answers is not in the document]
Comment (ovidiuolteanu) [reply contracts/fees_collector_contract.py:55]: Legacy stuff, mostly irrelevant now when forcing `no_init = True`.

**Review by claudiulataretu — APPROVED** [review]

**Pull request merged**